### PR TITLE
♻️ use MarketState subgraph entity 

### PR DESCRIPTION
**Pull request by sebipap**

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -12,11 +12,10 @@ if (!MARKET) throw new Error('missing market');
 console.log(MARKET);
 
 Promise.all([
-  queryRates(SUBGRAPH_URL, MARKET, 'deposit', { maxFuturePools: 3 }),
```
Comment (sebipap) [inline src/main.ts:15]: `maxFuturePools` prop is not needed anymore, it's derived from `MarketState`

File: src/api.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    maturity,
+    timestamp: poolTimestamp,
+    unassignedEarnings: BigInt(unassignedEarnings),
+    lastAccrual,
+  })) : null,
+});
+
+const query = (timestamp: number | null, market: string): string => `
+      query {
+        marketStates(
```
Comment (cruzdanilo) [inline src/api.ts]: fix identation

File: src/queryRates.ts [hunk 3]
```diff
@@ -1,273 +1,54 @@
...
-      && (BigInt(earningsAccumulator) * elapsed)
-        / (elapsed + (BigInt(smoothFactor) * BigInt(maxFuturePools * FIXED_INTERVAL)) / WAD))
-
-    + ((
-      totalFloatingBorrowAssets(timestamp, marketState, floatingDebtState, interestRateModel)
-      - BigInt(floatingDebt)
-    ) * (WAD - BigInt(treasuryFeeRate))) / WAD
-  );
+  return {
+    date: new Date(startTimestamp * 1000),
```
Comment (cruzdanilo) [inline src/queryRates.ts]: `1_000`

File: src/queryRates.ts [hunk 4]
```diff
@@ -1,273 +1,54 @@
...
-    interval?: number;
-    offset?: number;
-    count?: number;
+    roundTicks?: boolean,
+    interval?: number,
+    offset?: number,
+    count?: number,
   } = {},
-) => {
+): Promise<Rate[]> => {
```
Comment (cruzdanilo) [inline src/queryRates.ts]: revert

Comment (sebipap) [reply src/queryRates.ts]: why? Should I always avoid return types?

File: src/api.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+          where: {
+            timestamp_lte: ${timestamp}
+            market_: { id: "${market}" }
+          }
+          orderBy: timestamp
+          first: 1
+          orderDirection: desc
+
```
Comment (itofarina) [inline src/api.ts]:
```suggestion
```

File: src/api.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            borrowed
+            supplied
+            lastAccrual
+          }
+          totalSupply
+        }
+      }
+      `;
+
+export default async function fetchMarketState(
```
Comment (itofarina) [inline src/api.ts]: I'd suggest renaming the file to `fetchMarketState` and
```suggestion
export default async (
```

File: src/utils.ts [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+const shareValue = (type: 'deposit' | 'borrow', marketState: MarketState, timestamp: number) => {
+  const assets = (type === 'deposit' ? totalAssets : totalFloatingBorrowAssets)(timestamp, marketState);
+  const shares = type === 'deposit' ? marketState.totalSupply : marketState.totalFloatingBorrowShares;
+
+  if (shares === 0n) return 0n;
+  return (assets * WAD) / shares;
+};
+
+export const change = (
```
Comment (itofarina) [inline src/utils.ts]:
```suggestion
export const proportion = (
```

File: src/utils.ts [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+  currentMarketState: MarketState,
+  previousMarketState: MarketState,
```
Comment (itofarina) [inline src/utils.ts]:
```suggestion
  current: MarketState,
  previous: MarketState,
```

**Review by itofarina — COMMENTED** [review]
Some minor comments but looking great!
I'd suggest also splitting the `utils` module preferring default exports

File: src/queryRates.ts [hunk 9]
```diff
@@ -1,273 +1,54 @@
...
-  interestRateModel: InterestRateModel,
-) => {
-  const newUtilization = BigInt(floatingAssets) > 0n
-    ? (BigInt(floatingDebt) * WAD) / BigInt(floatingAssets)
-    : 0n;
-  const borrowRate = floatingRate(
-    interestRateModel,
-    min(BigInt(utilization), newUtilization),
-    max(BigInt(utilization), newUtilization),
+export const rate = async (
```
Comment (itofarina) [inline src/queryRates.ts]: no need to export it
```suggestion
const rate = async (
```

File: src/utils.ts (hunk 8 above)
Comment (sebipap) [reply src/utils.ts:109]: why? I'd agree if the function name included 'marketState' in some way, but I think it's unclear what 'current' and 'previous' mean, but let me know if there was another motivation.

**Pull request closed without merging**